[BEAM-4391] Example of distributed optimization

javdrher opened the pull request:
This PR contributes an example to the Python SDK which uses Beam for distributed optimization. The example distributes a set of items over a set of factories, which need to be optimized using standard scientific libraries (e.g., SciPy). In order to do so, the pipeline generates a grid of all valid mappings of items on factories, then optimizes the factories for all these mappings. 

A test file is also included, I tried to copy the coding style from the other examples.

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

It will help us expedite review of your Pull Request if you tag someone (e.g. `@username`) to look at it.

aaltay commented:
R: @davidcavazos (David, could you make a first pass review?)

@javdrher Thank you. Please add license text to all files.

davidcavazos commented on sdks/python/apache_beam/examples/complete/distribopt/distribopt/distribopt.py:
```diff
@@ -0,0 +1,350 @@
+#
```
Does this file really has to exist within two `distribopt` subdirectories? Since this is a single file (plus test, but not a dependency), it shouldn't necessarily require the whole `setup.py` file and packaged directory, does it?

javdrher replied:
The only purpose of this structure is to make sure scipy and numpy are installed. I believe numpy is installed by default on dataflow, but what about scipy? I could also have it in a directory and bundle it with a requirements.txt

javdrher commented on sdks/python/apache_beam/examples/complete/distribopt/distribopt/distribopt.py:
```diff
@@ -0,0 +1,350 @@
+        # include the incomplete mapping of 2 crops
+        mapping.update(element)
+        # include identifier
+        mappings.append((uuid.uuid4().hex, mapping))
```
Thanks, added that improvement in 1a8db7a .

davidcavazos commented:
LGTM

aaltay commented on sdks/python/apache_beam/examples/complete/distribopt/distribopt/distribopt.py:
```diff
@@ -0,0 +1,349 @@
+
+class CreateGrid(beam.PTransform):
+  """
+  A transform for generating the mapping grid.
```
This could be a single line pydoc.

aaltay commented on sdks/python/apache_beam/examples/complete/distribopt/distribopt/distribopt.py:
```diff
@@ -0,0 +1,349 @@
+
+    def process(self, element):
+      records = list(element[1])
+      # Split of 2 crops and pre-generate all combinations to facilitate parallellism
```
I do not understand this comment.

Could you change comments to reflect both the nature of the example and also the Beam concepts it is showing.

aaltay commented on sdks/python/apache_beam/examples/complete/distribopt/distribopt/distribopt.py:
```diff
@@ -0,0 +1,349 @@
+
+  class GenerateMappings(beam.DoFn):
+    """
+    ParDo implementation to generate all possible assignments of crops to greenhouses.
```
You can start pydoc comments in the first line with """. (For examples see: https://github.com/apache/beam/blob/master/sdks/python/apache_beam/examples/complete/tfidf.py#L52)

aaltay commented on sdks/python/apache_beam/examples/complete/distribopt/distribopt/distribopt.py:
```diff
@@ -0,0 +1,349 @@
+    mappings = (
+        o.splitted
+        | 'create mappings' >> beam.ParDo(self.GenerateMappings(), pvalue.AsSingleton(o.combine))
+        | 'prevent fusion' >> beam.Reshuffle()
```
Why are we preventing shuffle here?

It is hard to see why it is needed when pipeline is composed in multiple places. Consider moving pipeline construction into a single place.

aaltay commented on sdks/python/apache_beam/examples/complete/distribopt/distribopt/distribopt.py:
```diff
@@ -0,0 +1,349 @@
+
+class CreateTransportData(beam.DoFn):
+  """
+  Transform records to pvalues ((crop, greenhouse), transport_cost)
```
What does 'Transform records to pvalues' mean?

aaltay commented on sdks/python/apache_beam/examples/complete/distribopt/distribopt/distribopt.py:
```diff
@@ -0,0 +1,349 @@
+  """
+  Adds the transport cost for the crop to the production cost.
+
+  pvalues are of the form (crop, greenhouse, mapping, cost), the cost only
```
`elements` instead of `pvalues`.

aaltay commented on sdks/python/apache_beam/examples/complete/distribopt/distribopt/distribopt.py:
```diff
@@ -0,0 +1,349 @@
+  pipeline_options = PipelineOptions(pipeline_args)
+  pipeline_options.view_as(SetupOptions).save_main_session = True
+
+  with beam.Pipeline(options=pipeline_options) as pipe:
```
Please follow a similar naming to other examples. This would allow readers to recognize common parts easily. For example in this case you could use `p` instead of `pipe`.

javdrher commented:
Sorry for the delay, I was attending a conference last week.

@aaltay I addressed most comments I think, regarding the testing: I'm guessing numpy and scipy will not be added to the testing environment? From my side I can not easily drop the dependency from the code unfortunately. What is the recommended road to resolve this? Bringing the imports inline to the functions and mock them during testing?

Regarding the fusion: the grid generation is a prime example of an operation with a PCollection low number of elements (actually only 1) and a large output PCollection. The effect on the execution of the pipeline is exactly as described [here](https://cloud.google.com/dataflow/service/dataflow-service-desc#fusion-optimization), hence the transform to prevent fusion. 

In fact, because the step from 1 element to a huge number of elements was itself slow, I create the grid in two stages: the first step takes two parameters and forms a small grid only with the options for that parameter, the second step then takes the small grid and extends it based on the options for the remaining parameters. That way, some parallelism is involved in the construction of the grid itself. (Actually, I think you could generalize this approach and implement the entire generation recursively into the pipeline but that seemed to be a bit out of scope and perhaps a bit abusive?). 

To encapsulate the grid creation process, I put it in a PTransform. I did the same for the optimization part of the pipeline (The PTransforms represent two phases: generate all options, optimize all options). Were you suggesting I drop the PTransforms and put the construction of the pipeline in one function? If so, I'll adjust the code.

aaltay commented:
Thank you @javdrher. This looks much better, I don't have any more file level comments.

To answer your questions:
> Bringing the imports inline to the functions and mock them during testing?
Sounds like a good idea to me.

We recently added numpy to the setup.py (test section) for performance tests. If you can mock only scipy related parts it should be enough.

> Regarding the fusion:  ...
This explanation is grid, please add it to the example with a link to that blog post. It would serve as a good learning opportunity.

> Were you suggesting I drop the PTransforms and put the construction of the pipeline in one function? If so, I'll adjust the code.
I think the current organization looks clean. You can keep it as is.

I could not review the simulation parts of the code as I am not familiar with this simulation. If you have a colleague who is familiar please ask them to take a quick look at those parts.

Does it answer your questions? Do you have a reasonable path to make progress?

javdrher commented:
@aaltay yes, thank you very much. I updated the testing code with a scipy mock, and solved all pylint problems. 

Regarding the simulation: it does not reflect something real, its an [analytical function](https://www.sfu.ca/~ssurjano/hart3.html) which I modified slightly to serve the purpose.
